### PR TITLE
Documentation folder and design document of agents

### PR DESCRIPTION
**Pull request by mhailstone**
I've captured the content from a Google document where much of the discussion has been happening on agents and put it into the markdown document here for content to be tracked and changed over time.

### CODE REVIEW

File: docs/README.md [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+#### Other Cloud Agent Functions
+A Cloud Agent may have other functions beyond message transport. 
+
+A key feature might be providing Wallet backup and restore functionality - holding an encrypted version of an Edge Agent’s wallet. The encryption and encryption key handling would be on the Agent side, but the encrypted data could be stored with the Cloud Agent. The encryption key handling would involve all of the DKMS mechanisms described in the [Evernym/DHS document](https://github.com/hyperledger/indy-sdk/blob/677a0439487a1b7ce64c2e62671ed3e0079cc11f/doc/design/005-dkms/DKMS%20Design%20and%20Architecture%20V3.md). So many details to be added…
+
+#### Variations
+In an Institution environment, a Cloud Agent is less important since the Edge Agent is likely a persistent end point using on-premise or Cloud hosting. It may still be useful to have a Cloud Agent to prevent correlation. Regardless, an IEA without a Cloud Agent should still look to the rest of the world like any other Agent, so it should have endpoint message handling like that of a Cloud Agent - e.g. expect to be sent anon-crypt messages with a transport instructions and an encrypted payload.
```
Comment (dbluhm) [inline docs/README.md]: "expect to be sent anon-crypt messages with (*a* omitted) transport instructions and an encrypted payload" or some other wording? It reads a little funny at the moment.

File: docs/README.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+### Hubs
```
Comment (dbluhm) [inline docs/README.md]: I'm interested to hear what others in the community think about the idea of "Hubs" and including them in our agent definitions; mostly regarding the following line:

> the Hub also stores and shares data (potentially including Verifiable Credentials), on behalf of its owner.

Personally, I'm a little uncomfortable about the idea of a cloud agent/hub having control over aspects of someone's identity (Verifiable Credentials) but I definitely see cloud agents potentially having more functionality beyond just message relaying. For instance, acting as a remote storage solution or the example given here of an online music service.

In general, I think anything that isn't under an individual's direct control should not be trusted with anything more than providing a basic service. But, again, I'm interested in what others have to say on this.

Pinging @nage and @dhh1128 specifically for comment.

File: docs/README.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+The process of handling messages should be consistent and independent of the transport mechanism of an agent implementation. The payload put onto the transport should be created in the following way:
+* Anon-encrypt the base message structure
+* [Base64](https://tools.ietf.org/html/rfc4648) encode the encrypted message
+
+The point in the process after the transport mechanism has delivered the payload should also be consistent and follow these steps:
+
+* [Base64](https://tools.ietf.org/html/rfc4648) decode the string
+* Anon-decrypt the bytes
+
+In order to anon encrypt/decrypt the payload, the endpoint (transport) mechanism MUST use a well known DID/VerKey associated with that endpoint called an Endpoint DID. No matter what kind of transport is used, the Endpoint DID should be used to package/unpackage the payload through/from the transport by following the steps outlined above.
```
Comment (dbluhm) [inline docs/README.md]: Secure transport of messages from source agent to destination agent is possible without a DID written to the ledger.
I'm willing to sit down and talk this over again.

File: docs/README.md [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+  "type": "URN message type",
+  "message": {}
+}
+```
+* The `id` attribute is required and needs to be either the DID of the sender that is a pairwise identifier in an established connection, a nonce used in establishing a connection, or another similar identifier used in a different custom message exchange.
+* The `type` attribute is required and MUST be a recognized message type.
+* The `message` attribute is required or optional depending upon the value of the type attribute and would contain the contents specified in the definition of the message type.
+ 
+### Message Types
+Message types MUST be in the form of a URN. For example, Sovrin implementations of Indy agents follow the structure of the URN as:
```
Comment (dbluhm) [inline docs/README.md]: I may be mistaken but I thought that the exact form of message types had not yet been decided. If it has been, I think having a reference to that decision somewhere would be good to back up the use of an all capital "MUST."

File: docs/README.md (hunk 3 above)
Comment (mhailstone) [reply docs/README.md]: That's true. I'm using the phrase "well known" here to convey that you could either lookup the DID on the ledger, or share the Endpoint DIDs between two endpoints out of band. Either way, both agent endpoints need to "know" the Endpoint DID details in order to send messages back and forth.

File: docs/README.md (hunk 4 above)
Comment (mhailstone) [reply docs/README.md]: The discussion we had in the Friday, June 15, 2018 agent call had consensus to follow this outlined message protocol.

**Pull request merged**